File: backend/app/schemas/variance.py
"""Variance Analysis Schemas"""

from pydantic import BaseModel, Field


class VarianceAnalysisDetail(BaseModel):
    """Variance analysis between forecast and actuals."""

    forecast_id: str
    week_number: int = Field(..., ge=1, le=52)
    forecasted_cumulative: int = Field(..., ge=0, description="Cumulative forecasted units")
    actual_cumulative: int = Field(..., ge=0, description="Cumulative actual units sold")
    variance_pct: float = Field(..., ge=0.0, le=1.0, description="Variance percentage (0.0-1.0)")
    threshold_exceeded: bool = Field(..., description="True if variance >20%")
    action_taken: str = Field(..., description="'reforecast_triggered' or 'none'")

File: frontend/src/types/variance.ts
// Variance types for weekly performance tracking

export interface StoreVariance {
  store_id: string;
  store_name: string;
  forecasted: number;
  actual: number;
  variance_pct: number;
}

export interface WeeklyVariance {
  forecast_id: string;
  week_number: number;
  forecasted_cumulative: number;
  actual_cumulative: number;
  variance_pct: number;
  threshold_exceeded: boolean;
  action_taken: string | null;
  store_level_variance: StoreVariance[];
}
